go/src/pcap/pcap.go

```diff
@@ -120,47 +116,95 @@ var zMapMap map[uint16]map[int]int
 
 var masscanMap map[uint16]map[int]int
 
+var scanMap map[uint16]map[uint16]int
+var firstPacketTime map[uint16]time.Time
+var recentPacketTime map[uint16]time.Time
+
+//ip source to scan sizes
+var scansSizes map[uint16][]int
+
 // TODO: add map that counts unique ip destinations as well
 // This map counts port destinations, but not ip dests. need both to classify scans and scan size
 //var zMapMapConcurrent sync.Map
 var scanMut sync.Mutex
-
+var zmapMut sync.Mutex
+var massMut sync.Mutex
 var mut sync.Mutex
 
 /* ========================= Main Loop ========================== */
 
-func packetRateGood(packet1 gopacket.Packet, packet2 gopacket.Packet) bool {
-	// packetType := fmt.Sprintf("%T", packet1.Metadata().Timestamp)
-	//fmt.Println(packetType)
-
-	// Type: time.Time
-	start := packet1.Metadata().Timestamp
-	end := packet2.Metadata().Timestamp
-
-	difference := end.Sub(start)
-
-	//fmt.Printf("Diff: %v\n", difference)
-
-	goalDuration, err := time.ParseDuration("100ms")
+func packetRateCheck(recent time.Time, ipSrc uint16, ipDest uint16) {
+	scanMut.Lock()
+	previousPacket := recentPacketTime[ipSrc]
+	oldestPacket := firstPacketTime[ipSrc]
+	allDests := scanMap[ipSrc]
+	if previousPacket.IsZero() {
+		//first packet for this ipSource scan
+		recentPacketTime[ipSrc] = recent
+		firstPacketTime[ipSrc] = recent
+		newDestMap := make(map[uint16]int)
+		newDestMap[ipDest] = 1
+		scanMap[ipSrc] = newDestMap
+		scanMut.Unlock()
+		return
+	}
+	scanMut.Unlock()
+	difference := recent.Sub(previousPacket)
+	expireTime, err := time.ParseDuration("480s")
 	if err != err {
 		log.Fatal(err)
 	}
 
-	return difference < goalDuration
+	longDifference := int(recent.Sub(oldestPacket))
+	numPackets := 0
+	scanMut.Lock()
+	for _, v := range allDests {
+		numPackets += v
+	}
+	scanMut.Unlock()
+	average := float64(numPackets) / (float64(longDifference) / float64(10e9))
```

Comment:
What exactly is this calculating, unit-wise? the cutoff is 0.1, but what is that units for?

Reply:
that is packets per second. So numPackets is number of packets. longDifference is time difference in nanoseconds, so i divide it by 10e9